server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java

```diff
@@ -127,8 +130,10 @@ public String blobName(String name) {
     public T readBlob(BlobContainer blobContainer, String blobName) throws IOException {
         final BytesReference bytes = Streams.readFully(blobContainer.readBlob(blobName));
         final String resourceDesc = "ChecksumBlobStoreFormat.readBlob(blob=\"" + blobName + "\")";
-        try (ByteArrayIndexInput indexInput =
-                 new ByteArrayIndexInput(resourceDesc, BytesReference.toBytes(bytes))) {
+        try {
+            final IndexInput indexInput = bytes.length() > 0 ? new ByteBuffersIndexInput(
+                    new ByteBuffersDataInput(Arrays.asList(BytesReference.toByteBuffers(bytes))), resourceDesc)
+                    : new ByteArrayIndexInput(resourceDesc, BytesRef.EMPTY_BYTES);
```

Comment:
`ByteBuffersDataInput` throws if passed an empty list of buffers so we need this special case handling to throw consistent exceptions when the checksum fails because a blob was truncated to empty for some reason.